### PR TITLE
idempotent batch delta ingestion

### LINKED ISSUE #2099: Make hadoop delta ingestion retries easier
Hadoop delta ingestion retrying is problematic because it's not idempotent. It's possible for the job to push segments and insert them into the metadata store, but then fail after that for some reason (maybe some hardware died at an inopportune time). Currently, users desiring to retry delta ingestion tasks need to deal with this outside of Druid - possibly by checking the metadata store before retrying.

It would be good for Druid to help with automating this or some other method of doing retries.

### PR DESCRIPTION
**Pull request by himanshug**
give user the option to specify the segments for dataSource inputSpec,
so that they can do "idempotent" delta ingestion.

Fixes #2099

### CODE REVIEW

**Comment by himanshug** [comment]
@gianm I added a new coordinator endpoint to get list of used segments given a dataSource and list of intervals which looks like..
```
POST /druid/coordinator/v1/metadata/datasources/{dataSourceName}/segments

body - [interval1, interval2,...]
```

I don't like it being a POST api but can't think of a cleaner way of getting list of intervals in a GET based api. thoughts?

**Comment by gianm** [comment]
@himanshug I think POST is ok, it makes it a lot easier to deal with potentially large intervals lists that are too big for the url.

File: server/src/main/java/io/druid/server/http/MetadataResource.java [hunk 1]
```diff
@@ -135,6 +142,43 @@ public Object apply(DataSegment segment)
...
+
+    Response.ResponseBuilder builder = Response.status(Response.Status.OK);
+    if (full != null) {
+      return builder.entity(segments).build();
+    }
+
+    return builder.entity(
+        Iterables.transform(
+            segments,
+            new Function<DataSegment, Object>()
```
Comment (gianm) [inline server/src/main/java/io/druid/server/http/MetadataResource.java]: this could return a String, would make the code easier to read.

**Comment by gianm** [comment]
@himanshug this looks good to me, one issue remaining is that users could accidentally lose ingested data if the following happens,

1) process A starts a delta ingestion using this method, it fails
2) process B updates that interval with some new data
3) process A starts a delta ingestion using this method, it succeeds, wiping away data from (2)

or even,

1) process A reads list of segments
2) process B updates data for that interval
3) process A submits delta ingest job, wiping away data from (2)

This could be addressed using overlord locking, by having the hadoop task check on start that the segments in the ingestion spec actually correspond to the currently-used segments. Then the retry would be both idempotent and safe (it would protect against the race above).

if you prefer not to do this now, then I think this PR could go through as-is (since the method in this PR works OK if there is some external locking on the user's side) but we should file an issue to fix this for the case where the is no external locking.

File: server/src/main/java/io/druid/server/http/MetadataResource.java (hunk 1 above)
Comment (himanshug) [inline server/src/main/java/io/druid/server/http/MetadataResource.java]: changed

**Comment by himanshug** [comment]
@gianm added the check to be safe as well as idempotent.

File: docs/content/ingestion/update-existing-data.md [hunk 2]
```diff
@@ -75,8 +76,7 @@ For example
 
 #### `multi`
 
-This is a composing inputSpec to combine other inputSpecs. This inputSpec is used for delta ingestion. 
-Please note that delta ingestion is not an idempotent operation. We may add change things in future to make it idempotent.
+This is a composing inputSpec to combine other inputSpecs. This inputSpec is used for delta ingestion. Please note that you can have only one `dataSource` as child of `multi` inputSpec.
```
Comment (gianm) [inline docs/content/ingestion/update-existing-data.md]: I think we should document the process, it's probably too much to expect people to figure it out themselves.

1. Call the coordinator api to get the list of segments
2. Submit a "multi" task with a "dataSource" child that does include those segments
3. If you get a failure, go back to step 1 and repeat until you succeed

**Comment by gianm** [comment]
awesome @himanshug, will take a look.

File: indexing-hadoop/src/main/java/io/druid/indexer/HadoopIngestionSpec.java [hunk 3]
```diff
@@ -168,10 +169,33 @@ public static HadoopIngestionSpec updateSegmentListIfDatasourcePathSpecIsUsed(
...
+
+      if (ingestionSpecObj.getSegments() != null) {
+        //ensure that user supplied segment list matches with the segmentsList obtained from db
+        //this safety check lets users do test-n-set kind of batch delta ingestion where the delta
+        //ingestion task would only run if current state of the system is same as when they submitted
+        //the task.
+        List<DataSegment> userSuppliedSegmentsList = ingestionSpecObj.getSegments();
+
+        if (segmentsList.size() == userSuppliedSegmentsList.size()) {
+          Collections.sort(segmentsList);
```
Comment (gianm) [inline indexing-hadoop/src/main/java/io/druid/indexer/HadoopIngestionSpec.java]: this assumes that the arrays returned by segmentLister and contained in the ingestion spec obj will be mutable, which might not always be the case. would be more robust to make copies of them as HashSets or TreeSets and then compare the sets for equality.

File: indexing-hadoop/src/main/java/io/druid/indexer/HadoopIngestionSpec.java [hunk 4]
```diff
@@ -168,10 +169,33 @@ public static HadoopIngestionSpec updateSegmentListIfDatasourcePathSpecIsUsed(
...
+          Collections.sort(userSuppliedSegmentsList);
+
+          for (int i = 0; i < segmentsList.size(); i++) {
+            if (!segmentsList.get(i).equals(userSuppliedSegmentsList.get(i))) {
+              throw new IOException("user supplied segments list did not match with segments list obtained from db");
```
Comment (gianm) [inline indexing-hadoop/src/main/java/io/druid/indexer/HadoopIngestionSpec.java]: would be nice to have a test verifying this behavior.

**Comment by gianm** [comment]
@himanshug looks good, had a couple comments.

File: docs/content/ingestion/update-existing-data.md (hunk 2 above)
Comment (himanshug) [inline docs/content/ingestion/update-existing-data.md]: updated doc to be more explicit.

File: indexing-hadoop/src/main/java/io/druid/indexer/HadoopIngestionSpec.java (hunk 3 above)
Comment (himanshug) [inline indexing-hadoop/src/main/java/io/druid/indexer/HadoopIngestionSpec.java]: ok, making a copy in a hashset

File: indexing-hadoop/src/main/java/io/druid/indexer/HadoopIngestionSpec.java (hunk 4 above)
Comment (himanshug) [inline indexing-hadoop/src/main/java/io/druid/indexer/HadoopIngestionSpec.java]: added new tests

**Comment by himanshug** [comment]
@gianm all comments addressed

**Comment by gianm** [comment]
awesome :+1:

File: docs/content/design/coordinator.md [hunk 5]
```diff
@@ -105,6 +105,14 @@ Returns a list of all segments for a datasource as stored in the metadata store.
 
 Returns a list of all segments for a datasource with the full segment metadata as stored in the metadata store.
 
+* POST `/druid/coordinator/v1/metadata/datasources/{dataSourceName}/segments`
+Body: [interval1, interval2,...] for example ["2012-01-01T00:00:00.000/2012-01-03T00:00:00.000", "2012-01-05T00:00:00.000/2012-01-07T00:00:00.000"]
+Returns a list of all segments, overlapping with any of given intervals,  for a datasource as stored in the metadata store.
```
Comment (fjy) [inline docs/content/design/coordinator.md]: add space

File: docs/content/design/coordinator.md [hunk 6]
```diff
@@ -105,6 +105,14 @@ Returns a list of all segments for a datasource as stored in the metadata store.
...
+
+* POST `/druid/coordinator/v1/metadata/datasources/{dataSourceName}/segments?full`
+Body: [interval1, interval2,...] for example ["2012-01-01T00:00:00.000/2012-01-03T00:00:00.000", "2012-01-05T00:00:00.000/2012-01-07T00:00:00.000"]
```
Comment (fjy) [inline docs/content/design/coordinator.md]: add space, please follow same format as other descs

**Pull request merged**